python/src/equistore/operations/_utils.py

```diff
@@ -48,58 +48,121 @@ def _check_blocks(a: TensorBlock, b: TensorBlock, props: List[str], fname: str):
+                 ``'components'``. To check only if the ``'parameters'`` are consistent
+                 use ``props=None``, using ``'parameters'`` is allowed
+                  even though deprecated.
+    """
+    err_msg = f"Inputs to {fname} should have the same gradient:\n"
+    grads_a = a.gradients_list()
+    grads_b = b.gradients_list()
+
+    if len(grads_a) != len(grads_b) or (
+        not np.all([parameter in grads_b for parameter in grads_a])
+    ):
+        raise ValueError(f"Inputs to {fname} should have the same gradient parameters.")
+
+    if props is not None:
+        for parameter, grad_a in a.gradients():
+            grad_b = b.gradient(parameter)
+            for prop in props:
+                err_msg_len = (
+                    f'gradient ("{parameter}") {prop} of the two `TensorBlock` '
+                    "have not the same lenght."
+                )
+
+                err_msg_1 = (
+                    f'gradient ("{parameter}") {prop} are not the same or not in the '
+                    "same order."
+                )
+
+                err_msg_names = (
+                    f'gradient ("{parameter}") {prop} names are not the same '
+                    "or not in the same order."
+                )
+
+                if prop == "samples":
+                    if not len(grad_a.samples) == len(grad_b.samples):
+                        raise ValueError(err_msg + err_msg_len)
+                    if not grad_a.samples.names == grad_b.samples.names:
+                        raise ValueError(err_msg + err_msg_names)
+                    if not np.all(grad_a.samples == grad_b.samples):
+                        raise ValueError(err_msg + err_msg_1)
+                elif prop == "properties":
+                    if not len(grad_a.properties) == len(grad_b.properties):
+                        raise ValueError(err_msg + err_msg_len)
+                    if not grad_a.properties.names == grad_b.properties.names:
+                        raise ValueError(err_msg + err_msg_names)
+                    if not np.all(grad_a.properties == grad_b.properties):
+                        raise ValueError(err_msg + err_msg_1)
+                elif prop == "components":
+                    if len(grad_a.components) != len(grad_b.components):
+                        raise ValueError(err_msg + err_msg_len)
+
+                    for c1, c2 in zip(grad_a.components, grad_b.components):
+                        if not (c1.names == c2.names):
+                            raise ValueError(err_msg + err_msg_names)
+
+                        if not np.all(c1 == c2):
+                            raise ValueError(err_msg + err_msg_1)
+                elif prop == "parameters":
+                    # parameters are already checked at the begining but i want to give
+                    # the opportunity to the 'user' to use it, without problems
+                    continue
+                else:
```

Comment:
I think we can go for an `!=` here and leave the docstring.

```suggestion
                elif prop != "parameters":
                    # parameters are already checked at the begining but i want to give
                    # the opportunity to the 'user' to use it, without problems
```